runtime/lua/vim/filetype.lua

```diff
@@ -1741,8 +1963,15 @@ local pattern = {
   ['.*%.properties_.._.._.*'] = starsetf('jproperties'),
   ['.*%.vhdl_[0-9].*'] = starsetf('vhdl'),
   ['.*/%.fvwm/.*'] = starsetf('fvwm'),
+  ['.*fvwmrc.*'] = function(path, bufnr)
+    vim.b[bufnr].fvwm_version = 1
+    starsetf('fvwm')
```

Comment:
Hm, `starsetf` doesn't work this way. It returns a function which is itself meant to be used by `filetype.lua`.

We should add a second optional function argument to `starsetf` to allow it to "do something" as part of the function it returns.

Reply:
Hmm... To be honest, I'm not at all a fan of how `filetype.vim` mixes setting `filetype` and "flavor variables"; this makes it harder to expose `vim.filetype.match` for other purposes.

Maybe we should instead refactor so that all patterns return a string (the filetype) and optionally a callback that is then run by `filetype.match` (unless called with an option to skip it)? This will be a bit tricky for some of the `detect.lua` patterns, though...

Reply:
> We should add a second optional function argument to `starsetf` to allow it to "do something" as part of the function it returns.
I went with your suggestion for now since it's easiest to implement. Can still refactor later :)

Reply:
@gpanders I think your proposed solution will not always work since sometimes the `starsetf` function is called multiple times (with different arguments) within one pattern function. What do you think of the approach I came up with? :D

Reply:
Let's take a step back: the `starsetf` function really does just two things: 1) check the file path against `g:ft_ignore_pat`, and 2) set the priority of the match to `-math.huge`.

Those two things should be generalizable. So here's what I have in mind: modify `starsetf` to accept either a string or a function. If a string, then it acts exactly as it does right now: it performs steps (1) and (2) above and simply returns the given string as the filetype. If a function, then it still performs steps (1) and (2) but also evaluates the given function to retrieve the filetype.

@clason's point is separate, but still relevant as we will need to incorporate that solution into `starsetf` as well. But I think we can do that in a follow up PR (I have an idea on how to address this).

Here's a patch:

```diff
diff --git a/runtime/lua/vim/filetype.lua b/runtime/lua/vim/filetype.lua
index f5c29639c..ff6cdcef9 100644
--- a/runtime/lua/vim/filetype.lua
+++ b/runtime/lua/vim/filetype.lua
@@ -5,14 +5,15 @@ local M = {}
 ---@private
 local function starsetf(ft)
   return {
-    function(path)
+    function(path, bufnr)
+      local f = type(ft) == 'function' and ft(path, bufnr) or ft
       if not vim.g.ft_ignore_pat then
-        return ft
+        return f
       end
 
       local re = vim.regex(vim.g.ft_ignore_pat)
-      if re:match_str(path) then
-        return ft
+      if not re:match_str(path) then
+        return f
       end
     end,
     {
```

Then this can be used in one of two ways:

The current method:

```lua
['foo'] = starsetf('fooscript'),
```

or using a function:

```lua
['bar'] = starsetf(function(path, bufnr)
  if some_condition then
    return 'barscript'
  end
  return 'somethingelse'
end),
```

I think this should cover all the cases. I will make a follow up PR to address the "setting buffer variables" issue once we get this one merged. What do you think?

Reply:
That makes a lot of sense, thanks for the patch! By the way, I also added an optional `opts` table so it's possible to override the default priority of -infinity.